### ISSUE TITLE
Exercises filtering is very limited due to Firestore constraint

### ISSUE THREAD

**Issue by Doan-Pham** [Doan-Pham/Stren#7]
### TLDR 
In ExercisesScreen, user can filter exercises by name/exercise category/muscles trained by exercises. The filtering has 3 parts using Firestore methods. The last 2 parts can't happen together (Firestore's constraint). Can only let Firestore do first 2 filtering then application does the final filtering locally. However, application uses pagination so the dataset after first 2 filtering won't cover the whole database, which means a lot of entries that may satisfy the 3rd filtering are not returned

### Description
The filtering query has 3 parts, each implemented with respective methods provided by Firestore:
1. Check if exercise name is in the filtering range (Using
**whereGreaterThanOrEqualTo()** and **whereLessThanOrEqualTo())**
2. Check if exercise's category (a string) is IN range provided by a
List parameter (Using **whereIn()**)
3. Check if exercise's primary muscles (a string array) CONTAINS ANY
value in a range provided by a List parameter (Using
**whereArrayContainsAny()**)

### Problem
The methods in step 2 and step 3 can't be used at the same time for a query according to Firestore documentation. So the filtering needs to be broken down further: Use only 1 of the 2 methods and let the backend (Firestore) do the 1st filtering - Then the application does the 2nd filtering itself upon receiving data.

### Problem 2
However, this solution still has a drawback. Since the application uses pagination, only a small amount of data is loaded for each query (more will be loaded when user needs it). This means the dataset returned after the filtering on backend won't be  complete and when the application's local filtering happens on this incomplete dataset, a lot of supposedly valid entries won't be included

Example:
- Database:
{"Name A", "Category A", ["Muscle Z, Muscle X, Muscle Y]"}
{"Name A", "Category B", ["Muscle Z, Muscle X, Muscle Y]"}
{"Name A", "Category B", ["Muscle A, Muscle B, Muscle C]"}

- Query Filtering:
    a. Name - A; (On backend)
    b. Category: In [A, B];  (On backend)
    c. Muscle: Contains any of [A, B, C].  (Locally on application)
- Pagination - Dataset Limit: 2

=> The (a.) and (b.) filtering returns the first 2 entries (since
dataset limit is 2), then the final filtering removes both, since they
don't satisfy the (c.) condition. However, the 3rd entry satisfies all,
but wasn't included due to the pagination's dataset limit

### Options
1.  Remove pagination:
-> The dataset returned after filtering on backend will be complete and
all valid data is shown
-> High cost, since have to read the whole database

2. Optimized pagination usage
-> Only use pagination when there's no filtering to do.
-> Complicated to setup

3 (Current). Leave it as is
-> The incomplete dataset is a problem only when the dataset being read have too few entries that satisfy the final filtering. Otherwise, if there are enough entries, the user can just scroll to load more
-> Very random and unpredictable
